query/worker.go

```diff
@@ -0,0 +1,240 @@
+package query
+
+import (
+	"errors"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+var (
+	// ErrQueryTimeout is an error returned if the worker doesn't respond
+	// with a valid response to the request within the timeout.
+	ErrQueryTimeout = errors.New("did not get response before timeout")
+
+	// ErrPeerDisconnected is returned if the worker's peer disconnect
+	// before the query has been answered.
+	ErrPeerDisconnected = errors.New("peer disconnected")
+
+	// ErrJobCanceled is returned if the job is canceled before the query
+	// has been answered.
+	ErrJobCanceled = errors.New("job canceled")
+)
+
+// queryJob is the internal struct that wraps the Query to work on, in
+// addition to some information about the query.
+type queryJob struct {
+	index      uint64
+	timeout    time.Duration
+	encoding   wire.MessageEncoding
+	cancelChan <-chan struct{}
+	*Request
+}
+
+// queryJob should satisfy the Task interface in order to be sorted by the
+// workQueue.
+var _ Task = (*queryJob)(nil)
+
+// Index returns the queryJob's index within the work queue.
+//
+// NOTE: Part of the Task interface.
+func (q *queryJob) Index() uint64 {
+	return q.index
+}
+
+// jobResult is the final result of the worker's handling of the queryJob.
+type jobResult struct {
+	job  *queryJob
+	peer Peer
+	err  error
+}
+
+// worker is responsible for polling work from its work queue, and hand it to
+// the associated peer. It validates incoming responses with the current
+// query's response handler, and poll more work for the peer when it has
+// successfully received a response to the request.
+type worker struct {
+	// nextJob is a channel of queries to be distributed, where the worker
+	// will poll new work from.
+	nextJob chan *queryJob
+}
+
+// NewWorker creates a new worker.
+func NewWorker() *worker {
```

Comment:
Don't think that's a problem as long as it is adhering to the exported interface.

Reply:
IIRC, it'll force callers to declare the type up front in areas where it isn't passed in as an argument with a type of the interface.

Reply:
I changed it to use the exported type, since it makes it nicer to match the method signature in `blockManagerCfg` in later commits: https://github.com/lightninglabs/neutrino/pull/179/commits/814cf6b9479fdba3bb384eaf76e9b4f198d75525